### PR TITLE
Puf kill switches

### PR DESCRIPTION
**Pull request by puf**
Fixes https://github.com/firebase/FirebaseUI-Android/issues/20. But
also changes https://github.com/firebase/FirebaseUI-Android/issues/9 to
needing to override both `populdateView()` and `parseSnapshot()`.

### CODE REVIEW

File: library/src/main/java/com/firebase/ui/auth/core/AuthProviderType.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+import com.firebase.ui.R;
+import com.firebase.ui.auth.facebook.FacebookAuthProvider;
+import com.firebase.ui.auth.google.GoogleAuthProvider;
+import com.firebase.ui.auth.password.PasswordAuthProvider;
+import com.firebase.ui.auth.twitter.TwitterAuthProvider;
+
+import java.lang.reflect.InvocationTargetException;
+
+public enum AuthProviderType {
+    GOOGLE  ("google",   "com.firebase.ui.auth.google.GoogleAuthProvider",     R.id.google_button),
```
Comment (abehaskins) [inline library/src/main/java/com/firebase/ui/auth/core/AuthProviderType.java]: Can we not do just "google.GoogleAuthProvider" and assume the "com.firebase.ui.auth" in the constructor? I feel like that might read better

File: library/src/main/java/com/firebase/ui/auth/core/AuthProviderType.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    public String getName() {
+        return mName;
+    }
+    public int getButtonId() {
+        return mButtonId;
+    }
+
+    public FirebaseAuthProvider createProvider(Context context, Firebase ref, TokenAuthHandler handler) {
+        try {
+            Class<? extends FirebaseAuthProvider> clazz = (Class<? extends FirebaseAuthProvider>) Class.forName(mProviderName);
```
Comment (abehaskins) [inline library/src/main/java/com/firebase/ui/auth/core/AuthProviderType.java]: `clazz` makes me kinda cringe, is this something people do? In Javascript we'd save class_ but I have no idea what the standard in other languages is.

File: library/src/main/java/com/firebase/ui/auth/google/GoogleAuthProvider.java [hunk 3]
```diff
@@ -93,7 +82,7 @@ public void onResult(Status status) {
     public void login() {
         if (mGoogleApiClient.isConnected()) {
             Intent signInIntent = Auth.GoogleSignInApi.getSignInIntent(mGoogleApiClient);
-            mActivity.startActivityForResult(signInIntent, GoogleActions.SIGN_IN);
+            ((Activity)getContext()).startActivityForResult(signInIntent, GoogleActions.SIGN_IN);
```
Comment (abehaskins) [inline library/src/main/java/com/firebase/ui/auth/google/GoogleAuthProvider.java]: Curious about the reasoning for the change here, I'm assuming this functioned before, so why the change?

**Comment by abehaskins** [comment]
Overall lgtm - thanks for cleaning up a bit of my mess :)

File: library/src/main/java/com/firebase/ui/auth/core/AuthProviderType.java (hunk 1 above)
Comment (puf) [inline library/src/main/java/com/firebase/ui/auth/core/AuthProviderType.java]: Fixed

File: library/src/main/java/com/firebase/ui/auth/core/AuthProviderType.java (hunk 2 above)
Comment (puf) [inline library/src/main/java/com/firebase/ui/auth/core/AuthProviderType.java]: There is definitely a [convention to call these clazz](http://stackoverflow.com/questions/2529974/why-do-java-programmers-like-to-name-a-variable-clazz), but I'm open to any other name.

File: library/src/main/java/com/firebase/ui/auth/google/GoogleAuthProvider.java (hunk 3 above)
Comment (puf) [inline library/src/main/java/com/firebase/ui/auth/google/GoogleAuthProvider.java]: I pulled the mActivity member into the base class and gave that a
getContext() accessor method. That might have been in a previous refactor
already, with this change just getting rid of the mActivity member.

I am a bit torn on getContext(). Since we seem to most often want to use it
as an Activity, we might just keep it as an Activity and prevent the need
to downcast.
